### ISSUE TITLE
latest changes to not destroy/recreate quill on 'regeneration' breaks react-quill when changing modules (dirty property)

### ISSUE THREAD

**Issue by adamskwersky** [zenoamaro/react-quill#833]
Here is my sandbox

https://codesandbox.io/s/react-quill-template-u9c9c](https://codesandbox.io/s/react-quill-template-forked-r1eemr?file=/src/App.js)

To reproduce, 
Type anything in the editor, then click 'make read only'. Notice the editor disappears completely from the DOM.
Clicking again to make it writable doesn't bring the editor back.

The code in the sandbox is actually the demo with some modifications.

If you have a different set of modules depending on whether the quill is read only, you will see this bug.
It seems to a problem during the 'regeneration' because module is a dirty property. 
The new code that no longer deletes and recreates the editor seems to have broken this.
But the change that was that broke this seems to fix issues with react-mention module (I guess this makes sense since react-mention hooks into key presses.?)

You can flip the react-quill version between beta4 and 2.0.0 and you'll see in beta 4 it works, but in 2.0.0 it doesn't.

### Ticket due diligence
These no longer seem relevant:
- [] I have verified that the issue persists under ReactQuill `v2.0.0`
- [ ] I can't use the beta version for other reasons

### ReactQuill version

- [ ] master
- [ ] v2.0.0
- [ ] v2.0.0-beta.2
- [ ] v2.0.0-beta.1
- [ ] 1.3.5
- [ ] 1.3.4 or older
- [ ] Other (fork)

Not a bug in Quill as far as I can tell. This worked in beta 4.

**Comment by cmrd-senya** [comment]
I think this may be related to https://github.com/zenoamaro/react-quill/commit/20274fb7c3ee03577564f9dfb0e2f032930a0f0b#diff-0b5adbfe7b36e4ae2f479291e20152e33e940f7f265162d77f40f6bdb5da7405L297

**Comment by adamskwersky** [comment]
> I think this may be related to this change: [20274fb#diff-0b5adbfe7b36e4ae2f479291e20152e33e940f7f265162d77f40f6bdb5da7405L297](https://github.com/zenoamaro/react-quill/commit/20274fb7c3ee03577564f9dfb0e2f032930a0f0b#diff-0b5adbfe7b36e4ae2f479291e20152e33e940f7f265162d77f40f6bdb5da7405L297)
> 
> CC @g12i

Yes, that's the one. There is no more 'regeneration' happening. Just disconnecting and reconnecting the hook.
As for the double toolbar issue. I never faced it.

**Comment by adamskwersky** [comment]
> Would it be possible to produce a failing test case to help develop a solution that satisfies both use cases?

It's a great idea, and I can try to do it.

**Comment by adamskwersky** [comment]
> Would it be possible to produce a failing test case to help develop a solution that satisfies both use cases?

Trick question? Well I was able to write a test to show the latest code has the regenerate problem where quill editor disappears, but I cannot write a test that shows the double toolbar, because it requires the strict mode in React 18, and enzyme isn't supported for that version of React. Were you trying to get me to rewrite the tests with React Testing Library ? (wink)

IN any case, I think the issue is that the quill gets re-rendered twice at the start, so we are likely creating to Quill's. This was shown here:
https://github.com/zenoamaro/react-quill/issues/784#issuecomment-1106688663

I don't know how to push my changes so you can at least get the regenerate bug test. I guess you need me to fork and push there?

**Comment by adamskwersky** [comment]
I pushed my test to here:
https://github.com/adamskwersky/react-quill

**Comment by alexkrolick** [comment]
The legacy test suite is Enzyme, but the new assertions are in Cypress right?

**Comment by adamskwersky** [comment]
> The legacy test suite is Enzyme, but the new assertions are in Cypress right?
> 
> https://github.com/adamskwersky/react-quill/blob/master/test/index.spec.js
> 
> old suite runs here https://github.com/adamskwersky/react-quill/blob/master/test/index.spec.js#L195

Ok, well if you need it done, it'll have to wait a bit, Possibly in two weeks.
In the meantime I stuck with 2.0.0-beta4 and did get the quill-mentions to work. The trick is to import ReactQuill, {Quill} before importing the quill-mention (importing {Quill} here is crucial), and to make sure the callbacks used are functions defined somewhere and not anonymous ones added to the quill-mention properties.
